File: src/tafqeetj/converters/DecimalInWords.java
package tafqeetj.converters;

import java.util.Objects;

public class DecimalInWords {
    private IntegerInWords numberLeftDecimalPoint;
    private IntegerInWords numberRightDecimalPoint;
    private String fractionalRatioName;

    public void setNumberLeftDecimalPoint(IntegerInWords numberLeftDecimalPoint) {
        this.numberLeftDecimalPoint = numberLeftDecimalPoint;
    }

    public void setNumberRightDecimalPoint(IntegerInWords numberRightDecimalPoint) {
        this.numberRightDecimalPoint = numberRightDecimalPoint;
    }

    public void setFractionalRatioName(String decimalLengthName) {
        this.fractionalRatioName = decimalLengthName;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DecimalInWords that = (DecimalInWords) o;
        return numberLeftDecimalPoint.equals(that.numberLeftDecimalPoint)
                && numberRightDecimalPoint.equals(that.numberRightDecimalPoint)
                && Objects.equals(fractionalRatioName, that.fractionalRatioName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(numberLeftDecimalPoint, numberRightDecimalPoint, fractionalRatioName);
    }

    @Override
    public String toString() {
        StringBuilder numberInWords = new StringBuilder();
        numberInWords.append(numberLeftDecimalPoint.toString());
        numberInWords.append(" فاصلة ").append(numberRightDecimalPoint.toString());
        if (!numberRightDecimalPoint.equals(IntegerInWords.ZERO))
            numberInWords.append(" ").append(fractionalRatioName);
        return numberInWords.toString();
    }
}

File: src/main/kotlin/tech/moussa/tafqeetk/converters/DecimalInWords.kt
package tech.moussa.tafqeetk.converters

import java.util.*

class DecimalInWords {
    private var numberLeftDecimalPoint: IntegerInWords? = null
    private var numberRightDecimalPoint: IntegerInWords? = null
    private var fractionalRatioName: String? = null

    fun setNumberLeftDecimalPoint(numberLeftDecimalPoint: IntegerInWords?) {
        this.numberLeftDecimalPoint = numberLeftDecimalPoint
    }

    fun setNumberRightDecimalPoint(numberRightDecimalPoint: IntegerInWords?) {
        this.numberRightDecimalPoint = numberRightDecimalPoint
    }

    fun setFractionalRatioName(decimalLengthName: String?) {
        this.fractionalRatioName = decimalLengthName
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as DecimalInWords
        return (
            this.numberLeftDecimalPoint!! == that.numberLeftDecimalPoint &&
            this.numberRightDecimalPoint!! == that.numberRightDecimalPoint &&
            this.fractionalRatioName == that.fractionalRatioName
        )
    }

    override fun hashCode(): Int {
        return Objects.hash(this.numberLeftDecimalPoint, this.numberRightDecimalPoint, this.fractionalRatioName)
    }

    override fun toString(): String {
        val numberInWords = StringBuilder()
        numberInWords.append(this.numberLeftDecimalPoint.toString())
        // TODO: Localization for different languages
        numberInWords.append(" فاصلة ").append(this.numberRightDecimalPoint.toString())
        if (this.numberRightDecimalPoint!! != IntegerInWords.ZERO) {
            numberInWords.append(" ").append(this.fractionalRatioName)
        }
        return numberInWords.toString()
    }
}